cranelift/codegen/src/egraph/stores.rs

```diff
@@ -0,0 +1,263 @@
+                self.other = MemoryState::AfterInst(inst);
+            }
+        }
+    }
+
+    fn for_flags(&mut self, memflags: MemFlags) -> &mut MemoryState {
+        if memflags.heap() {
+            &mut self.heap
+        } else if memflags.table() {
+            &mut self.table
+        } else if memflags.vmctx() {
+            &mut self.vmctx
+        } else {
+            &mut self.other
+        }
+    }
+
+    fn meet_from(&mut self, other: &LastStores, loc: Inst) {
+        let meet = |a: MemoryState, b: MemoryState| -> MemoryState {
+            match (a, b) {
+                (a, b) if a == b => a,
+                _ => MemoryState::BeforeInst(loc),
+            }
+        };
+
+        self.heap = meet(self.heap, other.heap);
+        self.table = meet(self.table, other.table);
+        self.vmctx = meet(self.vmctx, other.vmctx);
+        self.other = meet(self.other, other.other);
+    }
+}
+
+/// An alias-analysis pass.
+pub struct AliasAnalysis {
+    /// Last-store instruction (or none) for a given load.
+    load_mem_state: FxHashMap<Inst, MemoryState>,
+}
+
+impl AliasAnalysis {
+    /// Perform an alias analysis pass.
+    pub fn new(func: &Function, cfg: &ControlFlowGraph) -> AliasAnalysis {
+        log::trace!("alias analysis: input is:\n{:?}", func);
+        let block_input = Self::compute_block_input_states(func, cfg);
+        let load_mem_state = Self::compute_load_last_stores(func, block_input);
+        AliasAnalysis { load_mem_state }
+    }
+
+    fn compute_block_input_states(
+        func: &Function,
+        cfg: &ControlFlowGraph,
+    ) -> SecondaryMap<Block, Option<LastStores>> {
+        let mut block_input = SecondaryMap::with_capacity(func.dfg.num_blocks());
+        let mut worklist: SmallVec<[Block; 8]> = smallvec![];
+        let mut worklist_set = FxHashSet::default();
+        let entry = func.layout.entry_block().unwrap();
+        worklist.push(entry);
+        worklist_set.insert(entry);
+        block_input[entry] = Some(LastStores::default());
+
+        while let Some(block) = worklist.pop() {
+            worklist_set.remove(&block);
+            let state = block_input[block].clone().unwrap();
+
+            trace!("alias analysis: input to {} is {:?}", block, state);
+
+            let state = func
+                .layout
+                .block_insts(block)
+                .fold(state, |mut state, inst| {
+                    state.update(func, inst);
+                    trace!("after {}: state is {:?}", inst, state);
+                    state
+                });
+
+            for succ in cfg.succ_iter(block) {
+                let succ_first_inst = func.layout.first_inst(succ).unwrap();
+                let succ_state = &mut block_input[succ];
+                let old = succ_state.clone();
+                succ_state
+                    .get_or_insert_with(|| LastStores::default())
+                    .meet_from(&state, succ_first_inst);
```

Comment:
I think this should be `.get_or_insert(state)`, because we want `meet_from` to leave the state unchanged if this is the first time we're setting it. Or just:

```suggestion
                if let Some(succ_state) = succ_state.as_mut() {
                    succ_state.meet_from(&state, succ_first_inst);
                } else {
                    *succ_state = state;
                }
```

Reply:
Ah, right; I still had the `Top` default state in my mind and had figured that meeting with an initial `Top` was harmless; but here the default is `Entry` so we need to special-case the first set.